### ISSUE TITLE
corrupted brick file generation

### ISSUE THREAD

**Issue by sbailey** [desihub/desispec#312]
This command generates a corrupted brick file:
```
quickbrick -b test2 --objtype BGS -n 100 \
--outdir temp \
--seed 784077896 --airmass 1.5 --verbose --zrange-bgs 0.1 0.2 \
--moon-phase 0.1 --moon-angle 30 --moon-zenith 20
```
Note: there is an open PR to merge quickbrick and quickgen to reduce identical code.  This was tested on desisim commit 620f28a344b58627e340e8603311eb80427cd97d.

Opening one of the `temp/brick-*-test2.fits` files has the warning:
```
WARNING: VerifyWarning: Error validating header for HDU #5 (note: Astropy uses zero-based indexing).
    'ascii' codec can't decode byte 0xa2 in position 82: ordinal not in range(128)
There may be extra bytes after the last HDU or the file is corrupted. [astropy.io.fits.hdu.hdulist]
```
The appears to be coming from the header reporting the size of the fibermap data (HDU4) to be different from what it actually is, causing it to start parting HDU5 at the wrong location.  I think this comes from `desispec.io.brick.Brick.add_objects` where the fibermap table and header are updated in-place, but I haven't gotten further than that.

**Comment by gdhungana** [comment]
This seems same but adding here with std error output when running quickbrick. This occured with previous unmerged quickbrick and now merged in quickgen also.

 quickgen --brickname "moon09" --objtype "ELG" --nspec 1000 --seed 1234 --moon-phase 0.9 --moon-angle 80 --moon-zenith 60 > moon09.log 2>&1

INFO:quickgen.py:275:main: Initializing SpecSim with config desi
INFO:quickgen.py:329:main: Generating ELG template
INFO:io.py:601:read_basis_templates: Reading /home/govinda/Desi/spectro/templates/basis_templates/v2.2/elg_templates_v2.0.fits
Traceback (most recent call last):
  File "/home/govinda/Desi/softwares/desisim/bin/quickgen", line 11, in <module>
    sys.exit(quickgen.main(args))
  File "/home/govinda/Desi/softwares/desisim/py/desisim/scripts/quickgen.py", line 644, in main
    fx = fits.open(filepath, mode='append')
  File "/usr/local/lib/python2.7/dist-packages/astropy/io/fits/hdu/hdulist.py", line 139, in fitsopen
    return HDUList.fromfile(name, mode, memmap, save_backup, cache, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/astropy/io/fits/hdu/hdulist.py", line 281, in fromfile
    save_backup=save_backup, cache=cache, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/astropy/io/fits/hdu/hdulist.py", line 839, in _readfrom
    hdu = _BaseHDU.readfrom(ffo, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/astropy/io/fits/hdu/base.py", line 423, in readfrom
    **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/astropy/io/fits/hdu/base.py", line 483, in _readfrom_internal
    header = Header.fromfile(data, endcard=not ignore_missing_end)
  File "/usr/local/lib/python2.7/dist-packages/astropy/io/fits/header.py", line 451, in fromfile
    padding)[1]
  File "/usr/local/lib/python2.7/dist-packages/astropy/io/fits/header.py", line 520, in _from_blocks
    raise IOError('Header missing END card.')
IOError: Header missing END card.

This only wrote b channel file, and crashed when opening it (to append the truth) giving IO error "Header missing END card." So the file seems to be corrupted, when first written.

However, all three channel brick files were successfully written with nspec 10 or 100 in the above command.

**Comment by gdhungana** [comment]
I dug up a little bit as this is stopping me to generate a correct file. @sbailey I think the issue is when adding the fibermap in the brick as a hdulist, as you pointed. It seems tricky as to why adding hdulist generated from fits.convenience.table_to_hdu(data) truncates the file. While if only data is updated separately (instead of full HDU), it seems to work. See below for this example. I am dumping same functions from desispec.io.brick but only using relevant pieces for testing.

`import numpy as np
import astropy.io.fits as fits
from astropy import table
import desiutil,desispec
from desispec.io import empty_fibermap,Brick

filepath='./check_brick.fits'
nspec=1000
nwave=3000

#- Create empty brick
header = dict(BRICKNAM='brick', CHANNEL='r')
brick = Brick(filepath, mode='update', header=header)
hdu_list=brick.hdu_list

#- Update fibermap
fhdu=hdu_list['FIBERMAP']
fibermap=empty_fibermap(nspec)
data2 = table.Table(fibermap)

if len(fhdu.data) > 0: #- append if data exists 
    orig_data = table.Table(fhdu.data)
    data2 = table.vstack([orig_data, data2])

#- unicode -> ascii columns
data2 = desiutil.io.encode_table(data2)
newhdu = fits.convenience.table_to_hdu(data2)
newhdu.header = fhdu.header
hdu_list['FIBERMAP'] = newhdu  #- Change here
hdu_list.writeto(filepath,clobber = True) #- write to the file
hdu_list.close()

#- append arbitraty flux
trueflux=np.ones((nspec,nwave))*0.75
header = desispec.io.util.fitsheader(header)
fx = fits.open(filepath, mode='append')  #- if file written was corrupted, Open gives IOError. 
fx.append(fits.ImageHDU(trueflux, name='_TRUEFLUX', header=header))
fx.flush()
fx.close()
`
This gives a truncation warning the first time and gives IOError("Header Missing END card") when updating. 

While in the line above where I have "#- Change here" comment, if I use instead,
`hdu_list['FIBERMAP'].data = newhdu.data`
This works okay writing file correctly. Header was updated to the old header in the first case also. 

Do we lose any information not updating the whole HDU, but updating data this way? Or there is something more subtle?

**Comment by sbailey** [comment]
I think that would be ok.  The fits file writer will take care of updating any required header keywords so that they match the shape of the data.  The astropy.io.fits (formerly pyfits) library is pretty fragile and 
I suspect that this kind of FITS file surgery is outside of its intended usage, so it would be best to add a test that specifically fails with the current code but passes with your new code, so that if this breaks again in the future with an astropy update we'll catch it quickly.  Brick creation will need a significant update to be able to scale up to the entire survey anyway, at which point we can revisit how updating a pre-existing brick should be done.

**Comment by fjaviersanchez** [comment]
That workaround worked for me. I also tried another workaround using `fits.open(os.path.join(path_in,filename),ignore_missing_end=True)` when reading the corrupt files but it generated a bunch of warnings. However, the data was fine.

**Comment by gdhungana** [comment]
@fjaviersanchez Yes at first I also set, ignore_missing_end=True, but realize that it will still be a corrupted file and everywhere it is opened, we have to have that boolean. I suspect if there are other consequences as that will still not be a correct fits file. It's probably better not to approve a bad fits file, with this boolean? 

I'll add a test and send a PR with the earlier workaround later today.

**Comment by sbailey** [comment]
fixed via PR #314; thanks @gdhungana
